File: Servo/Servo.CS/Servo.cs
using NLog;
using System;
using System.Collections;
using System.Collections.Generic;
using System.ComponentModel;
using System.Configuration;
using System.Configuration.Install;
using System.Linq;
using System.ServiceProcess;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace Servo
{
    #region Config
    class Config
    {
        public string ServiceName
        {
            get
            {
                const string key = "Servo/Conf/ServiceName";

                var val = ReadConf(key);
                if (string.IsNullOrWhiteSpace(val)) throw new SettingsPropertyNotFoundException("app setting " + key + " is not set");

                return val;
            }
        }
        public string DisplayName
        {
            get
            {
                const string key = "Servo/Conf/DisplayName";

                var val = ReadConf(key);
                if (string.IsNullOrWhiteSpace(val)) throw new SettingsPropertyNotFoundException("app setting " + key + " is not set");

                return val;
            }
        }
        public ServiceStartMode ServiceStartMode
        {
            get
            {
                const string key = "Servo/Conf/ServiceStartMode";

                var val = ReadConf(key);
                if (string.IsNullOrWhiteSpace(val)) throw new SettingsPropertyNotFoundException("app setting " + key + " is not set");

                ServiceStartMode mode;
                if (!System.Enum.TryParse(val, out mode)) throw new SettingsPropertyNotFoundException("app setting " + key + " has a wrong value");

                return mode;
            }
        }
        public string Description
        {
            get
            {
                const string key = "Servo/Conf/Description";

                var val = ReadConf(key);
                if (string.IsNullOrWhiteSpace(val)) throw new SettingsPropertyNotFoundException("app setting " + key + " is not set");

                return val;
            }
        }
        public string[] ServicesDependedOn
        {
            get
            {
                const string key = "Servo/Conf/ServicesDependedOn";

                var val = ReadConf(key);
                if (string.IsNullOrWhiteSpace(val)) return new string[] { };

                return val.Split(new char[] { ',' }, StringSplitOptions.RemoveEmptyEntries);
            }
        }
        public bool DelayedAutoStart
        {
            get
            {
                const string key = "Servo/Conf/DelayedAutoStart";

                var val = ReadConf(key);
                if (string.IsNullOrWhiteSpace(val)) throw new SettingsPropertyNotFoundException("app setting " + key + " is not set");

                bool auto;
                if (!bool.TryParse(val, out auto)) throw new SettingsPropertyNotFoundException("app setting " + key + " has a wrong value"); ;

                return auto;
            }
        }
        public bool RunInService
        {
            get
            {
                const string key = "Servo/Conf/RunInService";

                var val = ReadConf(key);
                if (string.IsNullOrWhiteSpace(val)) throw new SettingsPropertyNotFoundException("app setting " + key + " is not set");

                bool runInService;
                if (!bool.TryParse(val, out runInService)) throw new SettingsPropertyNotFoundException("app setting " + key + " has a wrong value"); ;

                return runInService;
            }
        }

        public override string ToString()
        {
            var depended = string.Join(", ", ServicesDependedOn);
            return new { ServiceName, DisplayName, ServiceStartMode, Description, ServicesDependedOn = depended, DelayedAutoStart, RunInService }.ToString();
        }

        static string ReadConf(string key)
        {
            return ConfigurationManager.AppSettings[key];
        }
    }
    #endregion

    #region IService
    interface IService
    {
        void OnContinue();

        void OnCustomCommand(int command);

        void OnPause();

        bool OnPowerEvent(PowerBroadcastStatus powerStatus);

        void OnSessionChange(SessionChangeDescription changeDescription);

        void OnShutdown();

        void OnStart(string[] args);

        void OnStop();
    }
    #endregion

    #region InService
    class InService : ServiceBase
    {
        IService _scaffold;
        public InService(IService scaffold)
        {
            _scaffold = scaffold;

            var conf = new Config();
            ServiceName = conf.ServiceName;

            CanHandlePowerEvent = false;
            CanHandleSessionChangeEvent = false;
            CanPauseAndContinue = false;
            CanShutdown = false;
            CanStop = true;
            AutoLog = true;
        }

        protected override void OnContinue()
        {
            try { _scaffold.OnContinue(); }
            catch (NotImplementedException) { base.OnContinue(); }
        }

        protected override void OnCustomCommand(int command)
        {
            try { _scaffold.OnCustomCommand(command); }
            catch (NotImplementedException) { base.OnCustomCommand(command); }
        }

        protected override void OnPause()
        {
            try { _scaffold.OnPause(); }
            catch (NotImplementedException) { base.OnPause(); }
        }

        protected override bool OnPowerEvent(PowerBroadcastStatus powerStatus)
        {
            try { return _scaffold.OnPowerEvent(powerStatus); }
            catch (NotImplementedException) { return base.OnPowerEvent(powerStatus); }
        }

        protected override void OnSessionChange(SessionChangeDescription changeDescription)
        {
            try { _scaffold.OnSessionChange(changeDescription); }
            catch (NotImplementedException) { base.OnSessionChange(changeDescription); }
        }

        protected override void OnShutdown()
        {
            try { _scaffold.OnShutdown(); }
            catch (NotImplementedException) { base.OnShutdown(); }
        }

        protected override void OnStart(string[] args)
        {
            try { _scaffold.OnStart(args); }
            catch (NotImplementedException) { base.OnStart(args); }
        }

        protected override void OnStop()
        {
            try { _scaffold.OnStop(); }
            catch (NotImplementedException) { base.OnStop(); }
        }

        public void Run()
        {
            Run(this);
        }
    }
    #endregion

    #region InApp
    class InApp
    {
        IService _scaffold;
        MainForm _form;
        public InApp(IService scaffold)
        {
            _scaffold = scaffold;
            _form = new MainForm(_scaffold);
        }

        public void Run()
        {
            Application.Run(_form);
        }
    }
    #endregion

    #region MainForm
    class MainForm : Form
    {
        IService _svc;

        public MainForm(IService svc)
        {
            InitializeComponent();

            _svc = svc;
            Start();
        }

        private async void stopBtn_Click(object sender, EventArgs e)
        {
            await Stop();
        }

        async void Start()
        {
            stopBtn.Enabled = false;
            await Task.Factory.StartNew(() => _svc.OnStart(Environment.GetCommandLineArgs()), TaskCreationOptions.LongRunning);
            stopBtn.Enabled = true;
        }

        async Task Stop()
        {
            await Task.Factory.StartNew(() => _svc.OnStop(), TaskCreationOptions.LongRunning);
            this.Close();
        }

        /// <summary>
        /// Required designer variable.
        /// </summary>
        private System.ComponentModel.IContainer components = null;

        /// <summary>
        /// Clean up any resources being used.
        /// </summary>
        /// <param name="disposing">true if managed resources should be disposed; otherwise, false.</param>
        protected override void Dispose(bool disposing)
        {
            if (disposing && (components != null))
            {
                components.Dispose();
            }
            base.Dispose(disposing);
        }

        #region Windows Form Designer generated code

        /// <summary>
        /// Required method for Designer support - do not modify
        /// the contents of this method with the code editor.
        /// </summary>
        private void InitializeComponent()
        {
            this.stopBtn = new System.Windows.Forms.Button();
            this.SuspendLayout();
            // 
            // stopBtn
            // 
            this.stopBtn.FlatStyle = System.Windows.Forms.FlatStyle.Flat;
            this.stopBtn.Font = new System.Drawing.Font("Courier New", 12F, System.Drawing.FontStyle.Bold, System.Drawing.GraphicsUnit.Point, ((byte)(0)));
            this.stopBtn.Location = new System.Drawing.Point(12, 12);
            this.stopBtn.Name = "stopBtn";
            this.stopBtn.Size = new System.Drawing.Size(218, 72);
            this.stopBtn.TabIndex = 0;
            this.stopBtn.Text = "Stop";
            this.stopBtn.UseVisualStyleBackColor = true;
            this.stopBtn.Click += new System.EventHandler(this.stopBtn_Click);
            // 
            // MainForm
            // 
            this.AutoScaleDimensions = new System.Drawing.SizeF(6F, 13F);
            this.AutoScaleMode = System.Windows.Forms.AutoScaleMode.Font;
            this.ClientSize = new System.Drawing.Size(242, 96);
            this.Controls.Add(this.stopBtn);
            this.FormBorderStyle = System.Windows.Forms.FormBorderStyle.None;
            this.MaximizeBox = false;
            this.MinimizeBox = false;
            this.Name = "MainForm";
            this.ShowIcon = false;
            this.StartPosition = System.Windows.Forms.FormStartPosition.CenterScreen;
            this.Text = "Servo App";
            this.ResumeLayout(false);
        }

        #endregion

        private System.Windows.Forms.Button stopBtn;
    }
    #endregion

    #region SvcController
    class SvcController
    {
        readonly ServiceController _c = new ServiceController(new Config().ServiceName);

        public void Start()
        {
            try
            {
                ClassLogger.Info("Starting Service...");
                const int timeout = 7 * 60;
                _c.Start();

                const ServiceControllerStatus targetStatus = ServiceControllerStatus.Running;
                _c.WaitForStatus(targetStatus, TimeSpan.FromSeconds(timeout));
            }
            catch (Exception ex)
            {
                ClassLogger.Error("Error in starting service: {0}", ex);
            }
        }

        public void Stop()
        {
            try
            {
                ClassLogger.Info("Stopping Service...");
                const int timeout = 7 * 60;
                _c.Stop();

                const ServiceControllerStatus targetStatus = ServiceControllerStatus.Stopped;
                _c.WaitForStatus(targetStatus, TimeSpan.FromSeconds(timeout));
            }
            catch (Exception ex)
            {
                ClassLogger.Error("Error in stopping service: {0}", ex);
            }
        }

        static readonly Logger ClassLogger = LogManager.GetCurrentClassLogger();
    }
    #endregion

    #region SvcInstaller
    [RunInstaller(true)]
    public partial class SvcInstaller : Installer
    {
        public SvcInstaller()
        {
            var processInstaller = new ServiceProcessInstaller { Account = ServiceAccount.LocalSystem };

            var conf = new Config();
            var serviceInstaller = new ServiceInstaller
            {
                ServiceName = conf.ServiceName,
                DisplayName = conf.DisplayName,
                StartType = conf.ServiceStartMode,
                Description = conf.Description
            };

            try
            {
                serviceInstaller.ServicesDependedOn = conf.ServicesDependedOn;
            }
            catch (Exception ex) { ClassLogger.Error(ex); }

            try
            {
                serviceInstaller.DelayedAutoStart = conf.DelayedAutoStart;
            }
            catch (Exception ex) { ClassLogger.Error(ex); }

            Installers.AddRange(new Installer[] { serviceInstaller, processInstaller });
        }

        internal static void Install(bool undo, string[] args)
        {
            try
            {
                Console.WriteLine(undo ? "uninstalling" : "installing");

                if (undo)
                    ManagedInstallerClass.InstallHelper(new string[] { "/u", System.Reflection.Assembly.GetAssembly(typeof(SvcInstaller)).Location });
                else
                    ManagedInstallerClass.InstallHelper(new string[] { System.Reflection.Assembly.GetAssembly(typeof(SvcInstaller)).Location });


            }
            catch (Exception ex) { ClassLogger.Error(ex); }
        }

        static readonly Logger SlotClassLogger = LogManager.GetCurrentClassLogger();
        static Logger ClassLogger { get { return SlotClassLogger; } }
    }
    #endregion

    #region Extensions
    static class Extensions
    {
        public static void Run(this IService svc)
        {
            var conf = new Config();
            var flags = Environment.GetCommandLineArgs();
            var f = flags.Skip(1).FirstOrDefault() ?? string.Empty;

            if (f == "-h" || f == "-help")
            {
                ShowHelp(conf.RunInService);

                return;
            }

            if (conf.RunInService)
            {
                if (f == "-i" || f == "-install") SvcInstaller.Install(false, flags);
                else if (f == "-u" || f == "-uninstall") SvcInstaller.Install(true, flags);
                else if (f == "-start") new SvcController().Start();
                else if (f == "-stop") new SvcController().Stop();
                else if (f == "-a" || f == "-app") new InApp(svc).Run();
                else new InService(svc).Run();
            }
            else
            {
                new InApp(svc).Run();
            }
        }

        static void ShowHelp(bool runInService)
        {
            var h = new StringBuilder();
            if (!runInService) h = h.AppendLine("<< IS NOT RUNNING IN SERVICE MODE >>\r\nChange value of Servo/Conf/RunInService in App.config to true.\r\n");
            h = h
                .AppendLine("-h or -help\tshows this help")
                .AppendLine("-i or -install\tinstalls this windows service")
                .AppendLine("-u or -uninstall\tuninstalls this windows service")
                .AppendLine("-start\t\tstarts this windows service")
                .AppendLine("-stop\t\tstops this windows service")
                .AppendLine("-a or -app\t\truns this windows service as an app");

            MessageBox.Show(h.ToString(), "XService");
        }
    }
    #endregion
}

File: Servo/Servo.FS/Servo.fs
namespace Servo

open NLog
open System
open System.Collections
open System.Collections.Generic
open System.ComponentModel
open System.Configuration
open System.Configuration.Install
open System.Linq
open System.ServiceProcess
open System.Text
open System.Text.RegularExpressions
open System.Threading.Tasks
open System.Windows.Forms

module Helpers =
    let readConf (key: string) = ConfigurationManager.AppSettings.[key]

    let asEnum s :'a option when 'a:enum<'b> =
        match System.Enum.TryParse s with
        | true, v -> Some v
        | _ -> None

open Helpers

type Config() =
    member this.ServiceName with get() = let key = "WinSvc/Conf/ServiceName"
                                         let v = readConf key
                                         if String.IsNullOrWhiteSpace(v) then raise <| SettingsPropertyNotFoundException("app setting " + key + " is not set")
                                         v

    member this.DisplayName with get() = let key = "WinSvc/Conf/DisplayName"
                                         let v = readConf key
                                         if String.IsNullOrWhiteSpace(v) then raise <| SettingsPropertyNotFoundException("app setting " + key + " is not set")
                                         v

    member this.ServiceStartMode with get() = let key = "WinSvc/Conf/ServiceStartMode"
                                              let v = readConf key
                                              if String.IsNullOrWhiteSpace(v) then raise <| SettingsPropertyNotFoundException("app setting " + key + " is not set")
                                              
                                              let mode = (asEnum v : ServiceStartMode option)
                                              match mode with
                                              | Some(m) -> m
                                              | _ -> raise <| SettingsPropertyNotFoundException("app setting " + key + " has a wrong value")

    member this.Description with get() = let key = "WinSvc/Conf/Description"
                                         let v = readConf key
                                         if String.IsNullOrWhiteSpace(v) then raise <| SettingsPropertyNotFoundException("app setting " + key + " is not set")
                                         v

    member this.ServicesDependedOn with get() = let key = "WinSvc/Conf/ServicesDependedOn"
                                                let v = readConf key
                                                if String.IsNullOrWhiteSpace(v) then 
                                                    [||]
                                                else
                                                    v.Split([| ',' |], StringSplitOptions.RemoveEmptyEntries)

    member this.DelayedAutoStart with get() = let key = "WinSvc/Conf/DelayedAutoStart"
                                              let v = readConf key
                                              if String.IsNullOrWhiteSpace(v) then raise <| SettingsPropertyNotFoundException("app setting " + key + " is not set")

                                              let ok, mode = Boolean.TryParse(v)
                                              if not ok then raise <| SettingsPropertyNotFoundException("app setting " + key + " has a wrong value")
                                              
                                              mode

    member this.RunInService with get() = let key = "Servo/Conf/RunInService"
                                          let v = readConf key
                                          if String.IsNullOrWhiteSpace(v) then raise <| SettingsPropertyNotFoundException("app setting " + key + " is not set")

                                          let ok, runInService = Boolean.TryParse(v)
                                          if not ok then raise <| SettingsPropertyNotFoundException("app setting " + key + " has a wrong value")
                                          
                                          runInService

    override this.ToString() =
        let depended = String.Join(", ", this.ServicesDependedOn)
        (("ServiceName", this.ServiceName),
         ("DisplayName", this.DisplayName),
         ("ServiceStartMode", this.ServiceStartMode),
         ("Description", this.Description),
         ("ServicesDependedOn", depended),
         ("DelayedAutoStart", this.DelayedAutoStart),
         ("RunInService", this.RunInService)).ToString()

type IService = 
    abstract member OnContinue      : unit -> unit
    abstract member OnCustomCommand : int -> unit
    abstract member OnPause         : unit -> unit
    abstract member OnPowerEvent    : PowerBroadcastStatus -> bool
    abstract member OnSessionChange : SessionChangeDescription -> unit
    abstract member OnShutdown      : unit -> unit
    abstract member OnStart         : string[] -> unit
    abstract member OnStop          : unit -> unit

type InService(scaffold: IService) =
    inherit ServiceBase()

    static let ClassLogger = LogManager.GetCurrentClassLogger()

    do
        let conf = Config()
        base.ServiceName <- conf.ServiceName

        base.CanHandlePowerEvent <- false
        base.CanHandleSessionChangeEvent <- false
        base.CanPauseAndContinue <- false
        base.CanShutdown <- false
        base.CanStop <- true
        base.AutoLog <- true

    override this.OnContinue() =
        try
            scaffold.OnContinue()
        with _ as NotImplementedException ->
            base.OnContinue()

    override this.OnCustomCommand(command: int) =
        try
            scaffold.OnCustomCommand(command)
        with _ as NotImplementedException ->
            base.OnCustomCommand(command)

    override this.OnPause() =
        try
            scaffold.OnPause()
        with _ as NotImplementedException ->
            base.OnPause()

    override this.OnPowerEvent(powerStatus: PowerBroadcastStatus) =
        try
            scaffold.OnPowerEvent(powerStatus)
        with _ as NotImplementedException ->
            base.OnPowerEvent(powerStatus)

    override this.OnSessionChange(changeDescription: SessionChangeDescription) =
        try
            scaffold.OnSessionChange(changeDescription)
        with _ as NotImplementedException ->
            base.OnSessionChange(changeDescription)

    override this.OnShutdown() =
        try
            scaffold.OnShutdown()
        with _ as NotImplementedException ->
            base.OnShutdown()

    override this.OnStart(args: string[]) =
        try
            scaffold.OnStart(args)
        with _ as NotImplementedException ->
            base.OnStart(args)

    override this.OnStop() =
        try
            scaffold.OnStop()
        with _ as NotImplementedException ->
            base.OnStop()

    member this.Run() = 
        ServiceBase.Run(this)

type MainForm(svc: IService) as thisObj =
    inherit Form()
    
    let mutable components : System.ComponentModel.IContainer = null
    let mutable stopBtn    : System.Windows.Forms.Button      = null

    do
        thisObj.InitializeComponent()
        thisObj.Start()

    member this.Start() =
        stopBtn.Enabled <- false
        Task.Factory.StartNew(fun () -> svc.OnStart(Environment.GetCommandLineArgs()), TaskCreationOptions.LongRunning).Wait()
        stopBtn.Enabled <- true

    member this.Stop() =
        stopBtn.Enabled <- false
        Task.Factory.StartNew(fun () -> svc.OnStop(), TaskCreationOptions.LongRunning).Wait()
        this.Close()

    member this.stopBtn_Click(sender: obj, e: EventArgs) =
        this.Stop()

    member this.InitializeComponent() =
        stopBtn <- new System.Windows.Forms.Button()
        this.SuspendLayout()
        // 
        // stopBtn
        // 
        stopBtn.FlatStyle <- System.Windows.Forms.FlatStyle.Flat
        stopBtn.Font <- new System.Drawing.Font("Courier New", 12.0f, System.Drawing.FontStyle.Bold, System.Drawing.GraphicsUnit.Point, 0uy)
        stopBtn.Location <- new System.Drawing.Point(12, 12)
        stopBtn.Name <- "stopBtn"
        stopBtn.Size <- new System.Drawing.Size(218, 72)
        stopBtn.TabIndex <- 0
        stopBtn.Text <- "Stop"
        stopBtn.UseVisualStyleBackColor <- true
        stopBtn.Click.AddHandler(new System.EventHandler(fun s e -> this.stopBtn_Click(s, e)))
        // 
        // MainForm
        // 
        this.AutoScaleDimensions <- new System.Drawing.SizeF(6.0f, 13.0f)
        this.AutoScaleMode <- System.Windows.Forms.AutoScaleMode.Font
        this.ClientSize <- new System.Drawing.Size(242, 96)
        this.Controls.Add(stopBtn)
        this.FormBorderStyle <- System.Windows.Forms.FormBorderStyle.None
        this.MaximizeBox <- false
        this.MinimizeBox <- false
        this.Name <- "MainForm"
        this.ShowIcon <- false
        this.StartPosition <- System.Windows.Forms.FormStartPosition.CenterScreen
        this.Text <- "WinSvc App"
        this.ResumeLayout(false)

    override this.Dispose(disposing: bool) =
        if disposing && (components <> null) then
            components.Dispose()

        base.Dispose(disposing)

type InApp(scaffold: IService) =
    let form = new MainForm(scaffold)

    member this.Run() = 
        Application.Run(form)

type SvcController() =
    let _c = new ServiceController(Config().ServiceName)
    static let ClassLogger = LogManager.GetCurrentClassLogger()

    member this.Start() = 
        try
            ClassLogger.Info("Starting Service...")
            
            let timeout = 7.0 * 60.0
            _c.Start()

            let targetStatus = ServiceControllerStatus.Running
            _c.WaitForStatus(targetStatus, TimeSpan.FromSeconds(timeout))
        with ex ->
            ClassLogger.Error(String.Format("Error in starting service: {0}", ex))

    member this.Stop() = 
        try
            ClassLogger.Info("Stopping Service...")

            let timeout = 7.0 * 60.0
            _c.Stop()

            let targetStatus = ServiceControllerStatus.Stopped
            _c.WaitForStatus(targetStatus, TimeSpan.FromSeconds(timeout))
        with ex ->
            ClassLogger.Error("Error in stopping service: {0}", ex)

[<RunInstaller(true)>]
type public SvcInstaller() as thisObj =
    inherit Installer()

    static let ClassLogger = LogManager.GetCurrentClassLogger()

    do
        let processInstaller = new ServiceProcessInstaller()
        let serviceInstaller = new ServiceInstaller()

        processInstaller.Account <- ServiceAccount.LocalSystem 

        let conf = Config()
        serviceInstaller.ServiceName <- conf.ServiceName    
        serviceInstaller.DisplayName <- conf.DisplayName    
        serviceInstaller.StartType   <- conf.ServiceStartMode 
        serviceInstaller.Description <- conf.Description

        try
            serviceInstaller.ServicesDependedOn <- conf.ServicesDependedOn
        with ex ->
            ClassLogger.Error(ex)

        try
            serviceInstaller.DelayedAutoStart <- conf.DelayedAutoStart
        with ex ->
            ClassLogger.Error(ex)

        thisObj.Installers.Add(serviceInstaller) |> ignore
        thisObj.Installers.Add(processInstaller) |> ignore
        
    static member Install (undo: bool) (args: String[]) = 
        try
            ClassLogger.Info(if undo then "uninstalling" else "installing")

            if undo then
                ManagedInstallerClass.InstallHelper([| "/u"; System.Reflection.Assembly.GetAssembly(typeof<SvcInstaller>).Location |])
            else
                ManagedInstallerClass.InstallHelper([| System.Reflection.Assembly.GetAssembly(typeof<SvcInstaller>).Location |])
        with ex ->
            ClassLogger.Error(ex)

module Toolbox =
    let ShowHelp(runInService : bool) =
        let mutable h = StringBuilder()
        if (not runInService) then h <- h.AppendLine("<< IS NOT RUNNING IN SERVICE MODE >>\r\nChange value of Servo/Conf/RunInService in App.config to true.\r\n")
        h <- h
            .AppendLine("-h or -help\tshows this help")
            .AppendLine("-i or -install\tinstalls this windows service")
            .AppendLine("-u or -uninstall\tuninstalls this windows service")
            .AppendLine("-start\t\tstarts this windows service")
            .AppendLine("-stop\t\tstops this windows service")
            .AppendLine("-a or -app\t\truns this windows service as an app")

        MessageBox.Show(h.ToString(), "XService") |> ignore

    type IService with
        member this.Run () =
            let conf = new Config()
            let flags = Environment.GetCommandLineArgs()
            let mutable f = flags.Skip(1).FirstOrDefault()
            if String.IsNullOrWhiteSpace(f) then f <- String.Empty

            match conf.RunInService, f with
            | _, "-h" | _, "-help" -> ShowHelp(conf.RunInService)
            | true, _ ->
                match f with
                | "-i" | "-install" -> SvcInstaller.Install false flags
                | "-u" | "-uninstall" -> SvcInstaller.Install true flags
                | "-start" -> (new SvcController()).Start()
                | "-stop" -> (new SvcController()).Stop()
                | "-a" | "-app" -> (new InApp(this)).Run()
                | _ -> (new InService(this)).Run()
            | _ -> (new InApp(this)).Run()
                
